### ISSUE TITLE
Coverage tracking sometimes misses a line

### ISSUE THREAD

**Issue by Hixie** [dart-lang/sdk#53349]
When I run the coverage for package:rfw, it usually says all the testable lines are covered, but every now and then, one line is marked as not covered. This isn't just a flaky test (running different code sometimes), because if you add a line immediately before the uncovered line, you can check that that line is covered even when the next is not:

![](https://github.com/dart-lang/sdk/assets/551196/11a5abdc-463a-4626-986d-e916fce1b485)

`hit` (a global bool variable initialized to false) is true (i added `expect(hit, isTrue)` to the tests, and it's true when i expect it to be), and that line is marked as covered, the red line is the only non-covered line, and yet you can't not reach it if you reach the previous line without throwing.

**Comment by Hixie** [comment]
Or more than a line:

![](https://github.com/dart-lang/sdk/assets/551196/bade03a0-daf6-47b9-922e-c2081cee162a)

That's the same switch statement case, I don't know why this time the `case` lines are marked as uncovered. Nor do I know why the `throw` is uncovered; that's _definitely_ tested. A couple of other throws later in the file also got marked as uncovered that time.

To find this I just ran `flutter test --coverage` in a tight loop until the coverage results changed. While I'm sure that unpredictable tests could be blamed for such changes in some situations, in this case I'm confident the tests are definitely covering this code.

**Comment by srawlins** [comment]
Tentatively triaging to VM.

**Comment by mkustermann** [comment]
@liamappelbe could you look at this?

**Comment by liamappelbe** [comment]
I am able to repro this. `flutter test --coverage test/text_test.dart` has about a 2-3% flake rate for coverage (the test still passes when coverage flakes), and misses that specific line each time. I haven't been able to repro the second screenshot, where many more lines are missed.

I also tried commenting out all the cases in text_test that hit that line (the second half of the `'parseLibraryFile: references'` test cases), and the flake went away (1000 runs so far without a flake). So it seems like a pretty specific case.

@Hixie Do you have any guesses as to what might be special about this line or these tests that could cause the flake?

**Comment by Hixie** [comment]
No, I have no idea at all. I spent quite some time trying to chase down theories (like, maybe there's a test parsing the time of day? maybe there's a race condition? etc) but could never substantiate any of them. In the end I marked the line with a pointer to this issue and make the coverage-checking script just skip this line in the lcov.info file.

I haven't tried narrowing it down, that's probably the next step in debugging this. By which I mean, remove code in the lib/ directory, try again, remove tests, try again, etc. But with a 2% flake, that'll be incredibly slow work.

I've only seen the failure in the second screenshot once, and I ran the tests nonstop overnight after seeing it to see if I could repro it, so if that's part of the same flakiness pattern, it's extremely rare. It might have just been an unrelated fluke.

**Comment by mraleph** [comment]
I would not be surprised if this is caused by JIT optimization kicking in at slightly different moments and causing changes to  coverage through that.

**Comment by a-siva** [comment]
lowering the priority to P3 since the flake rate is low and we are unable to easily reproduce it.

**Comment by nielsenko** [comment]
Here is another example of oddly missing coverage hits https://coveralls.io/builds/65863369/source?filename=packages%2Fejson%2Flib%2Fsrc%2Fdecoding.dart

**Comment by liamappelbe** [comment]
@nielsenko That's an unrelated bug. This issue is about a rare flaky coverage miss. Your example is a consistent miss related to the new pattern language features (probably another example of [this bug](https://github.com/dart-lang/sdk/issues/53519)).

**Comment by Mr-Pepe** [comment]
Is this maybe related to https://github.com/dart-lang/coverage/issues/482? I wanted to raise that issue here again but wanted to first check whether it's a duplicate of this issue.

I was also experiencing flaky coverage results. I turned out that it depended on the order in which the tests were run. If a test that generated a large number of line hits was run first, coverage for subsequent tests would not be counted correctly. From what I can tell, the VM is providing a faulty coverage report via RPC.

**Comment by Hixie** [comment]
(If the test order matters, it is sometimes also because the test order changes what code runs, I've definitely run into that before. Not saying that's what you're seeing, of course.)

**Comment by Mr-Pepe** [comment]
The random test order is what surfaced the issue but I was able to reduce it to a minimal example (see the linked issue). My question is whether that would explain the observed behavior here or if I should open a separate issue.

**Comment by liamappelbe** [comment]
@Mr-Pepe It could be related, as both bugs seem related to the optimiser. Hard to know for sure, since this issue is such a rare flake. I haven't had a chance to look at https://github.com/dart-lang/coverage/issues/482 yet, but once I fix it I can check if it also fixes this one.

No need to file a separate issue, since it's already tracked by https://github.com/dart-lang/coverage/issues/482.

**Comment by jsroest** [comment]
I think the problem is still here.

<img width="700" height="664" alt="Image" src="https://github.com/user-attachments/assets/2fe00764-0524-4eae-90fc-c1d8343fdafd" />

I asked my AI agent to investigate and it came with this response, which seems plausibe:

> ❯ If I look at the coverage for @packages/service_desk_server/lib/src/endpoints/hooked/hooked_private_endpoint.dart, I see that a lot of the session parameters are marked as not tested. For example line 57, 63 and 78. While line 28 is marked as tested. This dropped the coverage significantly. Why is that and can this be fixed in the tests?                                                                                        
  ⎿  Read packages/service_desk_server/lib/src/endpoints/hooked/hooked_private_endpoint.dart (185 lines)                                                
                                                                                                                                                        
> ⏺ Let me look at the existing tests for this endpoint and the coverage report.
> 
> ⏺ Searched for 3 patterns, read 2 files (ctrl+o to expand)
>
> ⏺ The tests exist and cover all those methods, so the issue is likely a stale coverage report. Let me check the coverage scripts and run fresh coverage to see what's actually happening.
>                                                                                                     
> ⏺ Searched for 2 patterns, read 1 file (ctrl+o to expand)
>
> ⏺ There are two sets of HTML files in coverage — let me check the script and the lcov data to understand what's happening.                              
>
> ⏺ Searched for 2 patterns, read 1 file (ctrl+o to expand)
>
> ⏺ Now I have the full picture. Let me read the raw lcov data alongside the source to map exactly what's happening.
>
> ⏺ Searched for 1 pattern (ctrl+o to expand)
>
> ⏺ Now I can give a clear diagnosis. Here's exactly what's happening:
>
> The tests DO execute those lines — it's a Dart VM coverage tool artifact, not missing test coverage.
>
> The evidence is in the lcov DA:line,count data:
>
```
Line  Content                                          Count
56    await SessionFisher.db.deleteWhere(              1 ← method IS called
57      session,                                       0 ← artifact
58      where: (row) => row...                         4 ← lambda runs 4×
60    );                                               0 ← artifact
62    await SessionFisher.db.insert(                   1 ← method IS called
63      session,                                       0 ← artifact
77    Future<void> deleteFishingSession(               1 ← method IS called
78      Session session, {                             0 ← artifact
81    await FishingSession.db.deleteRow(session, row)  1 ← executed
```
> 
>   The Dart VM places coverage markers at some individual-token positions (standalone positional arguments before named parameters, closing brackets, parameter declaration lines) but doesn't reliably fire them during execution. This is a well-known inconsistency in Dart's kernel-level coverage instrumentation.
> 
>  Notice the same pattern in createFishingSession — the session, on line 28 is not even tracked in the lcov file (it doesn't appear at all), so it  doesn't hurt the score there. The private endpoint just happens to generate more of these spurious markers.
>
>  Can it be fixed in tests? No — the tests already execute all the code. You can verify by running coverage fresh right now and checking that all the  method-call lines (56, 62, 77, 81) have counts ≥ 1.
>
>  Options to actually improve the score:
>
 > 1. Accept it. The 24 "uncovered" lines are all no-logic positions (argument identifiers, closing parens, parameter declarations).  hooked_public_endpoint.dart shows 100% precisely because its methods don't happen to generate those spurious markers.
>  2. Add // coverage:ignore-line on the affected lines — but that's noisy.
>  3. Report it — this is a known Dart issue (filed under the Dart SDK issue tracker, search "lcov coverage false zero").

**Comment by liamappelbe** [comment]
@jsroest To me that looks like you're applying a coverage report to the wrong source file, or maybe you (or your AI) changed or formatted the source code after the coverage report was generated. The lines that make me think that are 55 and 85. There's not even any source code on those lines, but they're reported as missing. We occasionally have bugs where non-executing code is marked as uncovered, but I've never seen a bug where whitespace is marked as uncovered.

If you're confident this is a real bug, can you create a minimal reproducible example? I have no way of investigating the issue otherwise.

Btw that AI response is slop. It's not an explanation, it's just restating the symptoms in more words. No idea what it means by "well-known inconsistency in Dart's kernel-level coverage instrumentation" either 😆.
